Store user model to simplify ONNXProgram.{adapt_torch_*,__call__} APIs

thiagocrepaldi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #115380
* #115353
* #114762
* __->__ #115281
* #114407

Currently (after https://github.com/pytorch/pytorch/pull/114407), the user has must pass the original user ``model`` to APIs such as ``ONNXProgram.__call__``, ``ONNXProgram.adapt_torch_inputs_to_onnx`` and ``ONNXProgram.adapt_torch_outputs_to_onnx`` APIs.

This was needed because when the model is fakefied, a version of the non-fakefied model is needed so that the Initializers, buffers and constants can be extracted from a real model (and used as input to the ONNX model).
That approach brings an unnecessary usability burden to the user when the model is not fakefied, because the model that was already passed to ``torch.onnx.dynamo_export`` could be used to extract ``state_dict``.

This PR adds ``ONNXProgram._model_torch`` attribute to store the user model and demote ``model`` argument of the aforementioned APIs to optional, only (as opposed to required).

As a result, for the fakefied model scenario, the user still need to pass the required model, but for non fakefied models, the persisted model is implicitly used to extract the model state_dict, making it easier to use.

Commit:
Store user model to simplify ONNXProgram.{adapt_torch_*,__call__} APIs

Currently, the user has must pass the original user ``model`` to
APIs such as ``ONNXProgram.__call__``,
``ONNXProgram.adapt_torch_inputs_to_onnx`` and
``ONNXProgram.adapt_torch_outputs_to_onnx`` APIs.

This is needed because when the model is fakefied, a version of the
non-fakefied model is needed so that the Initializers, buffers and
constants can be extracted (and used as input to the ONNX model).
That approach brings an unnecessary burden to the user when the model is
not fakefied, because the model that was already passed to
``torch.onnx.dynamo_export`` could be used to extract ``state_dict``

This PR adds ``ONNXProgram._model_torch`` to store the user model and
promotes the ``model`` argument to the aforementioned APIs to optional.
As a result, for the fakefied model scenario, the user can still pass
the required model, but for non fakefied models, the persisted model is
implicitly used to extract the model state_dict

[ghstack-poisoned]

Commit:
Update on "Store user model to simplify ONNXProgram.{adapt_torch_*,__call__} APIs"


Currently, the user has must pass the original user ``model`` to
APIs such as ``ONNXProgram.__call__``,
``ONNXProgram.adapt_torch_inputs_to_onnx`` and
``ONNXProgram.adapt_torch_outputs_to_onnx`` APIs.

This is needed because when the model is fakefied, a version of the
non-fakefied model is needed so that the Initializers, buffers and
constants can be extracted (and used as input to the ONNX model).
That approach brings an unnecessary burden to the user when the model is
not fakefied, because the model that was already passed to
``torch.onnx.dynamo_export`` could be used to extract ``state_dict``

This PR adds ``ONNXProgram._model_torch`` to store the user model and
promotes the ``model`` argument to the aforementioned APIs to optional.
As a result, for the fakefied model scenario, the user can still pass
the required model, but for non fakefied models, the persisted model is
implicitly used to extract the model state_dict

[ghstack-poisoned]

Commit:
Update on "Store user model to simplify ONNXProgram.{adapt_torch_*,__call__} APIs"


Currently, the user has must pass the original user ``model`` to
APIs such as ``ONNXProgram.__call__``,
``ONNXProgram.adapt_torch_inputs_to_onnx`` and
``ONNXProgram.adapt_torch_outputs_to_onnx`` APIs.

This is needed because when the model is fakefied, a version of the
non-fakefied model is needed so that the Initializers, buffers and
constants can be extracted (and used as input to the ONNX model).
That approach brings an unnecessary burden to the user when the model is
not fakefied, because the model that was already passed to
``torch.onnx.dynamo_export`` could be used to extract ``state_dict``

This PR adds ``ONNXProgram._model_torch`` to store the user model and
promotes the ``model`` argument to the aforementioned APIs to optional.
As a result, for the fakefied model scenario, the user can still pass
the required model, but for non fakefied models, the persisted model is
implicitly used to extract the model state_dict

[ghstack-poisoned]

BowenBao commented on test/onnx/onnx_test_common.py:
```diff
@@ -436,12 +436,15 @@ def _compare_pytorch_onnx_with_ort(
     ref_outputs = ref_model(*ref_input_args, **ref_input_kwargs)
-    ref_outputs = onnx_program.adapt_torch_outputs_to_onnx(ref_model, ref_outputs)
+    ref_outputs = onnx_program.adapt_torch_outputs_to_onnx(
+        ref_outputs, model_with_state_dict=ref_model
```
hmm why still need to pass ref_model when it is part of onnx_program init args?

thiagocrepaldi replied:
LOL From PR description "This is needed because when the model is fakefied, a version of the non-fakefied model is needed so that the Initializers, buffers and constants can be extracted (and used as input to the ONNX model)."

BowenBao replied:
still confused .. 2 lines above same `ref_model` is passed as init arg to onnx_program. How is that one fake while the one on L446 non fake

BowenBao replied:
wait ok that one is not init, but `__call__` ...

Commit:
Update on "Store user model to simplify ONNXProgram.{adapt_torch_*,__call__} APIs"


Currently, the user has must pass the original user ``model`` to
APIs such as ``ONNXProgram.__call__``,
``ONNXProgram.adapt_torch_inputs_to_onnx`` and
``ONNXProgram.adapt_torch_outputs_to_onnx`` APIs.

This is needed because when the model is fakefied, a version of the
non-fakefied model is needed so that the Initializers, buffers and
constants can be extracted (and used as input to the ONNX model).
That approach brings an unnecessary burden to the user when the model is
not fakefied, because the model that was already passed to
``torch.onnx.dynamo_export`` could be used to extract ``state_dict``

This PR adds ``ONNXProgram._model_torch`` to store the user model and
promotes the ``model`` argument to the aforementioned APIs to optional.
As a result, for the fakefied model scenario, the user can still pass
the required model, but for non fakefied models, the persisted model is
implicitly used to extract the model state_dict

[ghstack-poisoned]

Commit:
Update on "Store user model to simplify ONNXProgram.{adapt_torch_*,__call__} APIs"


Currently, the user has must pass the original user ``model`` to
APIs such as ``ONNXProgram.__call__``,
``ONNXProgram.adapt_torch_inputs_to_onnx`` and
``ONNXProgram.adapt_torch_outputs_to_onnx`` APIs.

This is needed because when the model is fakefied, a version of the
non-fakefied model is needed so that the Initializers, buffers and
constants can be extracted (and used as input to the ONNX model).
That approach brings an unnecessary burden to the user when the model is
not fakefied, because the model that was already passed to
``torch.onnx.dynamo_export`` could be used to extract ``state_dict``

This PR adds ``ONNXProgram._model_torch`` to store the user model and
promotes the ``model`` argument to the aforementioned APIs to optional.
As a result, for the fakefied model scenario, the user can still pass
the required model, but for non fakefied models, the persisted model is
implicitly used to extract the model state_dict

[ghstack-poisoned]

thiagocrepaldi commented on torch/onnx/_internal/exporter.py:
```diff
@@ -809,7 +823,7 @@ def fake_context(self) -> Optional[ONNXFakeContext]:
         self,
         *model_args,
-        model: Optional[
+        model_with_state_dict: Optional[
```
we need the `model`, because the lifted "constant tensors" are not part of the `ExportedProgram.state_dict` (?!?!?!)

I will investigate whether I can make that change on `ExportedProgram.state_dict` , simplifying the requirement to `state_dict` only (which is the ideal case)

Commit:
Update on "Store user model to simplify ONNXProgram.{adapt_torch_*,__call__} APIs"


Currently, the user has must pass the original user ``model`` to
APIs such as ``ONNXProgram.__call__``,
``ONNXProgram.adapt_torch_inputs_to_onnx`` and
``ONNXProgram.adapt_torch_outputs_to_onnx`` APIs.

This is needed because when the model is fakefied, a version of the
non-fakefied model is needed so that the Initializers, buffers and
constants can be extracted (and used as input to the ONNX model).
That approach brings an unnecessary burden to the user when the model is
not fakefied, because the model that was already passed to
``torch.onnx.dynamo_export`` could be used to extract ``state_dict``

This PR adds ``ONNXProgram._model_torch`` to store the user model and
promotes the ``model`` argument to the aforementioned APIs to optional.
As a result, for the fakefied model scenario, the user can still pass
the required model, but for non fakefied models, the persisted model is
implicitly used to extract the model state_dict

[ghstack-poisoned]

Commit:
Update on "Store user model to simplify ONNXProgram.{adapt_torch_*,__call__} APIs"


Currently, the user has must pass the original user ``model`` to
APIs such as ``ONNXProgram.__call__``,
``ONNXProgram.adapt_torch_inputs_to_onnx`` and
``ONNXProgram.adapt_torch_outputs_to_onnx`` APIs.

This is needed because when the model is fakefied, a version of the
non-fakefied model is needed so that the Initializers, buffers and
constants can be extracted (and used as input to the ONNX model).
That approach brings an unnecessary burden to the user when the model is
not fakefied, because the model that was already passed to
``torch.onnx.dynamo_export`` could be used to extract ``state_dict``

This PR adds ``ONNXProgram._model_torch`` to store the user model and
promotes the ``model`` argument to the aforementioned APIs to optional.
As a result, for the fakefied model scenario, the user can still pass
the required model, but for non fakefied models, the persisted model is
implicitly used to extract the model state_dict

[ghstack-poisoned]

Commit:
Update on "Store user model to simplify ONNXProgram.{adapt_torch_*,__call__} APIs"


Currently, the user has must pass the original user ``model`` to
APIs such as ``ONNXProgram.__call__``,
``ONNXProgram.adapt_torch_inputs_to_onnx`` and
``ONNXProgram.adapt_torch_outputs_to_onnx`` APIs.

This is needed because when the model is fakefied, a version of the
non-fakefied model is needed so that the Initializers, buffers and
constants can be extracted (and used as input to the ONNX model).
That approach brings an unnecessary burden to the user when the model is
not fakefied, because the model that was already passed to
``torch.onnx.dynamo_export`` could be used to extract ``state_dict``

This PR adds ``ONNXProgram._model_torch`` to store the user model and
promotes the ``model`` argument to the aforementioned APIs to optional.
As a result, for the fakefied model scenario, the user can still pass
the required model, but for non fakefied models, the persisted model is
implicitly used to extract the model state_dict

[ghstack-poisoned]

Commit:
Update on "Store user model to simplify ONNXProgram.{adapt_torch_*,__call__} APIs"


Currently (after https://github.com/pytorch/pytorch/pull/114407), the user has must pass the original user ``model`` to APIs such as ``ONNXProgram.__call__``, ``ONNXProgram.adapt_torch_inputs_to_onnx`` and ``ONNXProgram.adapt_torch_outputs_to_onnx`` APIs.

This was needed because when the model is fakefied, a version of the non-fakefied model is needed so that the Initializers, buffers and constants can be extracted from a real model (and used as input to the ONNX model).
That approach brings an unnecessary usability burden to the user when the model is not fakefied, because the model that was already passed to ``torch.onnx.dynamo_export`` could be used to extract ``state_dict``.

This PR adds ``ONNXProgram._model_torch`` attribute to store the user model and demote ``model`` argument of the aforementioned APIs to optional, only (as opposed to required).

As a result, for the fakefied model scenario, the user still need to pass the required model, but for non fakefied models, the persisted model is implicitly used to extract the model state_dict, making it easier to use.

[ghstack-poisoned]

BowenBao commented on torch/onnx/_internal/exporter.py:
```diff
@@ -659,6 +659,9 @@ class ONNXProgram:
     _fake_context: Final[Optional[ONNXFakeContext]]
     _export_exception: Final[Optional[Exception]]
     _model_signature: Final[Optional[torch.export.ExportGraphSignature]]
+    _model_torch: Final[
```
mark as experimental?

thiagocrepaldi replied:
Tracked by https://github.com/pytorch/pytorch/issues/115461

BowenBao reviewed (approved):
🚢 to unblock CI and experiments. Let's keep exploring and revisit https://github.com/pytorch/pytorch/issues/115461

thiagocrepaldi commented:
@pytorchbot merge
